MyThreadPool/MyThreadPool/MyThreadPool.cs

```diff
@@ -0,0 +1,211 @@
+﻿namespace MyThreadPool;
+
+using System.Collections.Concurrent;
+
+/// <summary>
+/// MyTPL.
+/// </summary>
+public class MyThreadPool
+{
+    private readonly CancellationTokenSource cts = new();
+    private int numberOfThreads;
+    private ConcurrentQueue<Action> actions = new();
+    private AutoResetEvent newTask = new(false);
+    private Thread[] threads;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MyThreadPool"/> class.
+    /// Constructor for thread pool.
+    /// </summary>
+    /// <param name="numberOfThreads">Number of threads to start.</param>
+    /// <exception cref="ArgumentOutOfRangeException"> Exception throws if the number of threads is negative.</exception>
+    public MyThreadPool(int numberOfThreads)
+    {
+        if (numberOfThreads <= 0)
+        {
+            throw new ArgumentOutOfRangeException("Number of threads must be positive");
+        }
+
+        this.numberOfThreads = numberOfThreads;
+        threads = new Thread[this.numberOfThreads];
+        for (var i = 0; i < threads.Length; i++)
+        {
+            threads[i] = new Thread(() =>
+            {
+                while (true)
+                {
+                    if (actions.TryDequeue(out Action? action))
+                    {
+                        action();
+                    }
+                    else
+                    {
+                        newTask.WaitOne();
```

Comment:
Мне казалось, что я поправил.
При добавлении задачи мы идем в EnqueueTask, складываем задачу и сеттим newTask, он должен будить рандомный поток, если 1 уже занят, то разбудит второй.

Reply:
Нет. ResetEvent --- это булевый флаг, он не знает, сколько раз был вызван EnqueueTask. Так что если в одном потоке EnqueueTask вызовется 10 раз подряд, а потом управление попадёт сюда, то тут проснётся только один поток. Никто не обещает, что сразу после Set проснётся поток на WaitOne, и между Set и WaitOne никто больше не сделает Set:
```
Important

There is no guarantee that every call to the Set method will release a thread from an EventWaitHandle whose 
reset mode is EventResetMode.AutoReset. If two calls are too close together, so that the second call occurs 
before a thread has been released, only one thread is released. It is as if the second call did not happen
```
(см. https://docs.microsoft.com/en-us/dotnet/api/system.threading.eventwaithandle.set?view=net-6.0#system-threading-eventwaithandle-set)